### ISSUE TITLE
Inconsistent number of records indexed using simple test

### ISSUE THREAD

**Issue by lukas-vlcek** [BitScoutOrg/efk-atomicapp#11]
Even after rsyslog allowed TCP (see https://github.com/BitScoutOrg/nulecule-rsyslog-collector/pull/2) I am experiencing inconsistent number of records indexed after simple test.
- We will use `make_rfc5424.py` script to generate artificial log records but we will modify it to produce more data. 
- We assume fresh and clean BitScout application running and Elasticsearch node available via HTTP.
- For maximal clarity of the test we will want to change default settings of BitScout Elasticsearch index.
### Start BitScout

See [quickstart-instructions](https://github.com/BitScoutOrg/efk-atomicapp/wiki/How-to-run-the-application-with-atomic#quickstart-instructions).
### Download and modify data creating script
- Download https://raw.githubusercontent.com/BitScoutOrg/efk-atomicapp/master/make-rfc5424.py
- We want to generate more data, change line [23](https://github.com/BitScoutOrg/efk-atomicapp/blob/master/make-rfc5424.py#L23) to `nfacility = 824`
- Now we have a script that will generate 6592 records:

```
$ python make_rfc5424.py | wc -l
6592
```
### Check the Elasticsearch cluster is empty

Assuming http://localhost:9200 points to HTTP endpoint of Elasticsearch node.

Point web browser to http://localhost:9200/_cat/indices?v. If the Elasticsearch is empty, you will get the following output:

:: PASTE SCREENSHOT HERE ::
### Create index template with custom settings

```
curl -X PUT 'localhost:9200/_template/bitscout_custom' -d '{
  "template": "bitscout*",
  "order": 20,
  "settings": {
    "refresh_interval"   : -1,
    "number_of_shards"   : 1,
    "number_of_replicas" : 0
  }
}'
```
### Index the data

Run the script that we have modified before:

```
python make_rfc5424.py | nc localhost 5141
```

Point web browser to http://localhost:9200/_cat/indices?v&h=health,status,index,pri,rep,docs.count,docs.deleted,store.size,pri.store.size,indexing.index_total,merges.current,merges.total

Output should list all existing indices. The most important is `indexing.index_total` column. It contains the real total number of indexed documents in this index. The column `docs.count` contains total number of documents in this index that are available via search API. In our case it will contain `0` due to previously set `refresh_interval` (or until the [flush](https://www.elastic.co/guide/en/elasticsearch/reference/1.5/indices-flush.html) operation is called automatically).

Index [refresh](https://www.elastic.co/guide/en/elasticsearch/reference/1.5/indices-refresh.html) is used to make indexed documents searchable and can be called manually via:

```
curl -X POST 'localhost:9200/_refresh'
```

Other useful resource to check can be http://localhost:9200/_cat/segments?v which lists physical segments of the index along with `searchable` and `docs.count` status. Basically the sum of searchable segments should equal to total number of searchable document in the index.
### What is the problem?

The problem is that `indexing.index_total` value does not seem to be predictable and it can differ from number of records generated by `make_rfc5424.py` (which in our case is `6592`).
- I experienced the following values `6577`, `6593`, `6594`, `6609`.
- Changing number of shards does not seem to have any effect.
### Repeate the test

First drop the index

```
curl -X DELETE 'localhost:9200/bitscout.2015.12.18'
```

or if you have only a single bitscout index, use:

```
curl -X DELETE 'localhost:9200/bitscout*'
```

and go back to **Index the data** chapter.

**Comment by richm** [comment]
Can we do a simple test with only elasticsearch?  That is, add and delete a bunch of records to elasticsearch repeatedly, to confirm that elasticsearch works as expected without rsyslog/qpid/fluentd?

**Comment by lukas-vlcek** [comment]
So we need to learn how exactly fluentd-elasticsearch integration pushes data to Elasticsearch (probably batch updates) so that we can use the same method in our tests.

**Comment by lukas-vlcek** [comment]
I think I have solid proof that log records are being stuffed somewhere between syslog and Elasticsearch. So we can leave out Elasticsearch and focus on rsyslog, qpid and fluentd instead.

Here is what I consider the proof:

The idea is to modify the data creating script to produce records having the same message for each individual batch. Despite the fact that we end up with different number of records indexed in each batch we can simply count records having the same message and in the long run we should get equal counts for each unique message value.

Assuming fresh and clean BitScout installation as discussed above. Let's modify the rest of the test a bit.

### Download and modify data creating script

Get original `make_rfc5424.py` file and modify only line [44](https://github.com/BitScoutOrg/efk-atomicapp/blob/master/make-rfc5424.py#L44) and put a simple number into message (starting with '1'). So the line will read: `'msg'   : '1'`.

The idea is to have identical message for all the records in particular batch.

We shall expect `192` records indexed in Elasticsearch now,all having the same message.

```
python make_rfc5424.py | wc -l
# 192
```

### Create index template with custom settings

Make sure the message field is not analyzed.

```
curl -X PUT 'localhost:9200/_template/bitscout_custom' -d '{
  "template": "bitscout*",
  "order": 20,
  "settings": {
    "refresh_interval"   : -1,
    "number_of_shards"   : 1,
    "number_of_replicas" : 0
  },
  "mappings": {
    "fluentd": {
      "properties": {
        "message": {
          "type": "string",
          "index": "not_analyzed"
        }
      }
    }
  }
}'
```

### Index the data

Delete the index first (if there is any)

```
curl -X DELETE 'localhost:9200/bitscout*'
```

Index the data now

```
python make_rfc5424.py | nc localhost 5141
```

Refresh the index after some time

```
curl -X POST 'localhost:9200/_refresh'
```

### Query aggregations about the data

Now, as the data is searchable, we can use aggregations to get some statistics

```
curl -X POST 'localhost:9200/_search?pretty' -d '{
  "size": 0,
  "query": {
    "match_all": {}
  },
  "aggs": {
    "message": {
      "terms": {
        "field": "message",
        "min_doc_count": 1,
        "size": 10000
      }
    }
  }
}' | more
```

Then we can simply delete Elasticsearch index

```
curl -X DELETE 'localhost:9200/bitscout*'
```

Modify the `make_rfc5424.py` and increase the counter in `msg` field (see above) and start a new indexing batch.

```
python make_rfc5424.py | nc localhost 5141
```

and again refresh the data, query for aggregations, so on...

## Outputs

Reminder: there are generated `192` records in every batch.

### 1

```
  "hits" : {
    "total" : 191,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "message" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [ {
        "key" : "1",
        "doc_count" : 190
      }, {
        "key" : "invalid or yet-unknown config file command 'SystemLogSocketName' - have you forgotten to load a module? [try http://www.rsyslog.com/e/3003 ]",
        "doc_count" : 1
      } ]
    }
  }
```

### 2

```
  "hits" : {
    "total" : 192,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "message" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [ {
        "key" : "2",
        "doc_count" : 190
      }, {
        "key" : "1",
        "doc_count" : 2
      } ]
    }
  }
```

### 3

```
  "hits" : {
    "total" : 192,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "message" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [ {
        "key" : "3",
        "doc_count" : 190
      }, {
        "key" : "2",
        "doc_count" : 2
      } ]
    }
  }
```

### 4

```
  "hits" : {
    "total" : 193,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "message" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [ {
        "key" : "4",
        "doc_count" : 191
      }, {
        "key" : "3",
        "doc_count" : 2
      } ]
    }
  }
```

### 5

```
  "hits" : {
    "total" : 190,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "message" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [ {
        "key" : "5",
        "doc_count" : 189
      }, {
        "key" : "4",
        "doc_count" : 1
      } ]
    }
  }
```

### 6

```
  "hits" : {
    "total" : 191,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "message" : {
      "doc_count_error_upper_bound" : 0,
      "sum_other_doc_count" : 0,
      "buckets" : [ {
        "key" : "6",
        "doc_count" : 188
      }, {
        "key" : "5",
        "doc_count" : 3
      } ]
    }
  }
```

## Conclusion

As we can see every script run can produce different number of records in Elasticsearch (both less and more than `192`) but counting sum of total records for each message value in long run we get exactly `192` records. So the problem seem to be stuffing of records somewhere before Elasticsearch.

**Comment by richm** [comment]
Looks like there can be some messages from rsyslog too, so we will have to remove those.  Even so, looks like 2 or 3 messages from the script are being dropped.

**Comment by lukas-vlcek** [comment]
I left the rsyslog error in output no.1 intentionally as it can point to possible config issue we might have here (though I am not sure it is related).

Actually, the messages are not being dropped, the output just demonstrates that they do not make it all to ES and a few of them stay somewhere else. But they usually make it into the ES later when subsequent batch of messages is generated and fed to processing pipe, but again some of messages from the subsequent batch stay on the way and do not make it into ES until the next batch is generated etc... It seems that the larger the batch is the larger is the variation of indexed messages per batch.

**Comment by lukas-vlcek** [comment]
I found that the messages are stuck on the rsyslog-collector container side (see below for "proof").

So far I do not know why exactly, but it can be result of how we use/setup the qpid-router & fluentd combo or some issue in proton integration plugins for rsyslog or fluentd parts.

As far as I understood (what I was told by messaging QE guys) qpid-router works just as a "proxy server" and can not persist the messages itself. It can deliver messages from producer to consumer unless for example the consumer signals that is can not consume more events, in which case the messages stays at the producer side waiting for consumer to signal free capacity to receive new messages.

Following this theory I tried to stop fluentd or rsyslog-collector container during my tests. I found out that if I get to the point where there were less records delivered to ES than how much was generated by the `make_rfc5424.py` script (meaning some messages were sitting somewhere in the transit route waiting) and I stopped the rsyslog container and started it again (meaning RAM data has been cleared) then I was always missing these records in the total sums later. Stopping and starting fluentd did not have any effect.

For example:

192 records were generated, but only 188 made it into ES. This means 4 records are waiting somewhere. If I stop fluentd and start again, and generate new 192 records then no matter how many makes it into ES this time we always receive some (or all) of those 4 records from the first batch. However, if I stop and start rsyslog-collector instead then I never see any of those missing records later.

So I think that for some reason qpid-router blocks retrieval from rsyslog which causes the messages to stay in rsyslog queue waiting. Why qpid-router blocks the retrieval is something we should concentrate on now. It might be that fluentd is unable to consume messages at this rate and signals to message bus that it can not receive more messages but fails to signal free capacity back later. Or something like that...

**Comment by richm** [comment]
Looks like kgiusti has something we can test with rsyslog instead of the "hack" we are using now: https://github.com/kgiusti/rsyslog/tree/amqp1-v7.4.7

**Comment by lukas-vlcek** [comment]
Do I understand it correctly kgiusti is working on new native amqp1 output plugin for rsyslog?
In essence this https://github.com/kgiusti/rsyslog/blob/amqp1-v7.4.7/plugins/omamqp1/omamqp1.c could replace https://github.com/BitScoutOrg/docker-rsyslog-collector/blob/master/omamqp1.py
Do I get it correctly?

**Comment by richm** [comment]
@lukas-vlcek yes, that is correct.

**Comment by lukas-vlcek** [comment]
So we probably need to talk to kgiusti to learn if there is a chance this will get to master and be released soon otherwise we can at least test this using his branch.

**Comment by richm** [comment]
I have built Ken Giusti's amqp plugin in rsyslog in a copr repo: https://copr.fedoraproject.org/coprs/rmeggins/rsyslog/

I've updated the bitscout/rsyslog-collector to use this. I've also fixed a bug in the bitscout/fluentd, and I've turned off any additional debug levels or logging in those containers.

One thing I have noticed is that it seems to be timing related.  That is, if I start the containers manually one by one, giving 10 seconds after starting each container, before running the test with make-rfc5424, I don't see any problems.  I always see 192 records added.  However, when using efk-atomicapp, the first time I run make-rfc5424, I see much fewer than 192 records.  When I run make-rfc5424 subsequent times, I see 192 additional records added every time.

Looking at nulecule, I don't see a way to specify any sort of wait time or any way to do interdependencies in the graph format in Nulecule.  I think we are going to have to change the individual containers, especially the qpid-proton one, so that e.g. `docker run` does not return until the enclosed daemon is ready to accept connections.  I'm not sure what that criteria is - just adding `sleep 10` is fragile.

**Comment by lukas-vlcek** [comment]
For the reference I tried testing the stack using [test](https://github.com/ViaQ/docker-rsyslog-collector/tree/test) branch of **docker-rsyslog-collector** via `integration-tests/simple-test.sh` but for some reason no messages made it into Elasticsearch. @richm note that I had to resolve couple of conflicts in the test branch (I forked on commit https://github.com/ViaQ/docker-rsyslog-collector/commit/b808b039e10d2078c512cd041e6579a48c1e40f7 and then did couple of cherry-picks from `master` while resolving conflicts, whole diff can be seen here: https://github.com/ViaQ/docker-rsyslog-collector/compare/test?expand=1). May be I screwed something up...

Though not sure if it is worth the effort to do this torture, it might be better to just wait for atomic app fix?

**Comment by richm** [comment]
yeah, not sure what's wrong.  Let's just wait and see if the nulecule guys can fix the bug.

**Comment by lukas-vlcek** [comment]
It seems that using C implementation of rsyslog amqp plugin fixes this issue! I tested it manually both using `integration-tests/simple-test.sh` and `atomic run` and I am getting consistent and expected results now.
I propose to close this issue once we fix #16 (which is currently blocking easy testing).

**Comment by richm** [comment]
https://github.com/ViaQ/efk-atomicapp/issues/16 is now fixed and closed
